Document systemcrypto, requirefips

dagood opened the pull request:
Add docs for:

* https://github.com/microsoft/go/pull/935 / https://github.com/microsoft/go/pull/950
* https://github.com/microsoft/go/pull/968

I tidied up a little, generally. I moved the detailed multi-GOEXPERIMENT into its own section (where explaining `-tags` fits too) and put the example build commands in collapsed sections to tighten up the section a little after adding the new examples.

Rendered: https://github.com/microsoft/go/blob/dev/dagood/doc-sys-req/eng/doc/fips/README.md

dagood committed:
Document systemcrypto, requirefips

jaredpar commented on eng/doc/fips/README.md:
```diff
@@ -37,59 +37,100 @@ It is important to note that an application built with Microsoft's Go toolchain
 
 The `GOEXPERIMENT` environment variable is used at build time to select a cryptographic library backend. This modifies the Go runtime included in the program to use the specified platform-provided cryptographic library whenever it calls a Go standard library crypto API. The `GOEXPERIMENT` values that pick a crypto backend are:
 
+- *Recommended:* `systemcrypto` automatically selects the crypto backend we suggest for the target platform
```
```suggestion
- *Recommended:* `systemcrypto` automatically selects the suggested crypto backend for the target platform
```

jaredpar replied:
Do we ever explicitly state that `systemcrypto` selects OpenSSL for Linux, CNG for windows, etc ... We give a few examples but there isn't really a full table that says what combo of values selects which crypto libraries. I was wondering if we should have something like 

|Target Platform| systemcrypto mapping | library |
| -- | -- | --| 
| Linux | opensslcrypto | OpenSSL |
| Windows | cngcrypto | Windows |

It's fairly implicit though so not sure if it's needed

dagood replied:
Hmm, yeah, the "we suggest" was meant to point to the Microsoft fork/project doing the suggesting (rather than upstream Go, which suggests using none of these 😄) but this isn't a great way to convey that. I like the idea of trying out a table. I think it would be good to mention that the mapping matches the internal crypto library usage policy requirements for Go. I'm thinking of even dropping the somewhat vague "recommend" term, which I think raises questions about why it's a recommendation (and is it *just* a recommendation?).

jaredpar commented on eng/doc/fips/README.md:
```diff
@@ -117,9 +158,25 @@ To make the Go runtime panic during program initialization if FIPS mode is not e
-## Usage: Extra runtime configuration options
+## Usage: Extra configuration options
+
+### Enable FIPS mode at compile time
```
Enable or requires?

dagood replied:
I struggled with this header. 😄 I started with "Require FIPS mode at compile time", but I see that as saying the *compilation environment* is what should run in FIPS mode. "Enable FIPS mode" seemed a little better because it uses the terms set up in the previous section ("enable FIPS mode" could mean "GOFIPS=1"), but it's not really any good, seeing it again now. I think this fixes it:

```suggestion
### Build option to require FIPS mode
```

jaredpar commented:
Now there are a lot of different options to control FIPS behavior:

- GOEXPERIMENT to select the backend 
- requiresfips build tag 
- GOFIPS environment variable

Do we think it's clear what happens for all the varations on these flags?

dagood reviewed (commented):
> Do we think it's clear what happens for all the varations on these flags?

Hmm, maybe not. I'll play with this. One thing in particular that I think will help is moving the "tutorial" info to a separate doc to keep these sections focused on the features we've added and how they interact.

dagood committed:
Update eng/doc/fips/README.md

dagood committed:
Update mentions of 1.22 to 1.21

dagood committed:
Add knob index and common config table

dagood committed:
"we suggest" -> "suggested", mention policy

dagood committed:
systemcrypto table, move policy note before table

dagood committed:
Merge 'microsoft/main' into dev/dagood/doc-sys-req

qmuntal reviewed (approved):
LGTM++

dagood committed:
Document allowcryptofallback

dagood committed:
Add dedicated changelog section

dagood committed:
Fix statement that GOFIPS=0 uses Go crypto

dagood committed:
Remove clickable detail/summary wrappers

Decided against after some time seeing them in this context. It's odd to need to click on them and they make the contents invisible to page search. They also don't save that much space after all.

dagood commented:
Last set of changes for sure this time. 😄 I want to get this in, then we can go back to smaller PRs.

I think having a dedicated changelog section is good--a simple place to put info about changes over time rather than just keeping notes around based on whether they seem useful or not. Anyone who might need to fix up a FIPS build after it sits for some amount of time could find it useful to get up to speed, too, especially if they don't follow our mailing list. (Or maybe they do, but this would be easier than looking back through to find what's relevant to building for FIPS.)

qmuntal commented on eng/doc/fips/README.md:
```diff
@@ -350,6 +352,36 @@ The work done to support FIPS compatibility mode leverages code and ideas from o
+
+- When multiple versions of OpenSSL are present on the machine at runtime, a version with FIPS mode enabled now has higher priority than others. [microsoft/go-crypto-openssl@v0.2.8](https://github.com/microsoft/go-crypto-openssl/releases/tag/v0.2.8)
+
+### Go 1.19 (Aug 2022)
```
Worth noting that the CNG backend was introduced in Go 1.19

dagood committed:
Include backend introductions in changelog

jaredpar commented on eng/doc/fips/README.md:
```diff
@@ -33,69 +31,165 @@ On Linux, the fork uses [OpenSSL](https://www.openssl.org/) through the [go-cryp
+| `GOEXPERIMENT=systemcrypto` | `GOFIPS=1` | Compliant | Can be used to create a compliant app. The app either enables FIPS mode or ensures it is already enabled. Otherwise, the app panics. |
+| `GOEXPERIMENT=systemcrypto` | `GOFIPS=0` | Compliant | Crypto usage is not FIPS compliant. The app attempts to disable FIPS mode and panics if it isn't possible. |
+| `GOEXPERIMENT=systemcrypto` | `GO_OPENSSL_VERSION_OVERRIDE=1.1.1k-fips` | Compliant | Can be used to create a compliant app. If the app is built for Linux, `systemcrypto` chooses `opensslcrypto`, and the environment variable causes it to load `libcrypto.so.1.1.1k-fips` instead of using the automatic search behavior. This environment variable has no effect with `cngcrypto`. |
+| `GOEXPERIMENT=systemcrypto` and `-tags=requirefips` | Default | Compliant | Can be used to create a compliant app. The behavior is the same as `GOFIPS=1`, but no runtime configuration is necessary. See [the `requirefips` section](#build-option-to-require-fips-mode) for more information on when this "locked-in" approach may be useful rather than the flexible approach. |
```
Think we should list the `allowcryptofallback` case here. Customers found it before, they'll find it again, best to acknowledge the combo exists and call it out as explicitly not compliant.

dagood replied:
It actually is compliant if used correctly, it just introduces risk. In any case, sure, adding.

dagood replied:
I put it into its own table at the end of this section instead to provide a little more context.

jaredpar commented on eng/doc/fips/README.md:
```diff
@@ -33,69 +31,165 @@ On Linux, the fork uses [OpenSSL](https://www.openssl.org/) through the [go-cryp
+  > Prior to Go 1.21, this alias is not available and the backend must be selected manually
 - `opensslcrypto` selects OpenSSL, for Linux
 - `cngcrypto` selects CNG, for Windows
 - `boringcrypto` selects the upstream BoringCrypto backend, which is **not supported**
```
Consider `not supported nor compliant`. Above we use compliant but here supported. Should enforce that it's not supported nor is it compliant

jaredpar commented on eng/doc/fips/README.md:
```diff
@@ -33,69 +31,165 @@ On Linux, the fork uses [OpenSSL](https://www.openssl.org/) through the [go-cryp
+| Target platform | `systemcrypto` selection | Library |
+| --- | --- | --- |
+| Linux | `opensslcrypto` | OpenSSL |
+| Windows | `cngcrypto` | CNG |
```
Consider adding a line for Mac stating it's not supported for completeness.

dagood committed:
PR feedback

More up-front allowcryptofallback example.

Be more specific that boringcrypto is not compliant.

Explicitly mention macOS is not supported with a link to a tracking issue.

dagood committed:
More precise systemcrypto build error condition
